File: src/db/sage/Kunde.ts
import { z } from "zod";
import { client, config } from "../config";

const Kunde = z.object({
  Name: z.string(),
  Vorname: z.string(),
});
type Kunde = z.infer<typeof Kunde>;

const getKunde = async (props: {
  Kundennummer: string;
}): Promise<Kunde | null> => {
  const form = new FormData();
  form.set("id", props.Kundennummer);
  const res = await client.post<Kunde | { error: string }>(
    "/Sage/Kunde",
    form,
    config
  );
  if ("error" in res.data) {
    return null;
  }
  return res.data;
};

const Sg_Adressen = z.object({
  SG_Adressen_PK: z.number().int(),
  Suchbegriff: z.object({ String: z.string(), Valid: z.boolean() }),
  KundNr: z.object({ String: z.string(), Valid: z.boolean() }),
  LiefNr: z.object({ String: z.string(), Valid: z.boolean() }),
  Homepage: z.object({ String: z.string(), Valid: z.boolean() }),
  Telefon1: z.object({ String: z.string(), Valid: z.boolean() }),
  Telefon2: z.object({ String: z.string(), Valid: z.boolean() }),
  Mobiltelefon1: z.object({ String: z.string(), Valid: z.boolean() }),
  Mobiltelefon2: z.object({ String: z.string(), Valid: z.boolean() }),
  EMail1: z.object({ String: z.string(), Valid: z.boolean() }),
  EMail2: z.object({ String: z.string(), Valid: z.boolean() }),
  KundUmsatz: z.object({ Valid: z.boolean(), Float64: z.number() }),
  LiefUmsatz: z.object({ Valid: z.boolean(), Float64: z.number() }),
});
type Sg_Adressen = z.infer<typeof Sg_Adressen>;

const SearchParams = z.object({
  Search: z.string(),
});
type SearchParams = z.infer<typeof SearchParams>;

const searchKunde = async (params: SearchParams): Promise<Sg_Adressen[]> => {
  const form = new FormData();
  form.set("Search", params.Search);
  const res = await client.post<Sg_Adressen[]>("/Sage/Search", form, config);
  return res.data;
};

const searchKundeReverse = async (
  params: SearchParams
): Promise<Sg_Adressen[]> => {
  const form = new FormData();
  form.set("Search", params.Search);
  const res = await client.post<Sg_Adressen[]>(
    "/Sage/ReverseSearch",
    form,
    config
  );
  return res.data;
};

export {
  getKunde,
  searchKunde,
  searchKundeReverse,
  type Kunde,
  SearchParams,
  type Sg_Adressen,
};

File: internal/handler/kunde.go
package handler

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log"
	"log/slog"
	"net/http"
	"regexp"
	"strings"

	"github.com/computerextra/golang-backend/internal/component"
	"github.com/computerextra/golang-backend/internal/types"
	"github.com/computerextra/golang-backend/internal/utils"

	_ "github.com/denisenkom/go-mssqldb"
)

func (h *Handler) SearchKunde(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	searchParams := r.URL.Query().Get("search")

	var SearchResults []types.Sg_Adressen

	if len(searchParams) < 1 {
		component.Kunde(SearchResults, "", false).Render(ctx, w)
		return
	}

	reverse, err := regexp.MatchString("^(\\d|[+]49)", searchParams)
	if err != nil {
		h.logger.Error("faulty regex", slog.Any("error", err))
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	connectionString, err := utils.GetSageConnectionString()
	if err != nil {
		h.logger.Error("failed to get connection string", slog.Any("error", err))
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	database, err := sql.Open("sqlserver", connectionString)
	if err != nil {
		h.logger.Error("failed to get sage connection", slog.Any("error", err))
		w.WriteHeader(http.StatusInternalServerError)
		return
	}
	defer database.Close()

	if reverse {
		SearchParam := strings.ReplaceAll(strings.ReplaceAll(strings.ReplaceAll(strings.ReplaceAll(strings.ReplaceAll(strings.ReplaceAll(strings.ReplaceAll(searchParams, "+49", "0"), " ", ""), "(", ""), ")", ""), "/", ""), "-", ""), ",", "")
		query := fmt.Sprintf(`
			SELECT SG_Adressen_PK, Suchbegriff,  KundNr, LiefNr, Homepage, Telefon1, Telefon2, Mobiltelefon1, Mobiltelefon2, EMail1, EMail2, KundUmsatz, LiefUmsatz 
			FROM sg_adressen WHERE 
			REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(Telefon1, ' ',''),'/',''),'-',''),'+49','0'),'(',''),')',''),',','')
			LIKE '%%%s%%' 
			OR 
			REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(Telefon2, ' ',''),'/',''),'-',''),'+49','0'),'(',''),')',''),',','')
			LIKE '%%%s%%' 
			OR 
			REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(Mobiltelefon1, ' ',''),'/',''),'-',''),'+49','0'),'(',''),')',''),',','')
			LIKE '%%%s%%' 
			OR 
			REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(Mobiltelefon2, ' ',''),'/',''),'-',''),'+49','0'),'(',''),')',''),',','')
			LIKE '%%%s%%'`, SearchParam, SearchParam, SearchParam, SearchParam,
		)

		rows, err := database.Query(query)
		if err != nil {
			h.logger.Error("failed to query sage", slog.Any("error", err))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
		defer rows.Close()

		for rows.Next() {
			var x types.Sg_Adressen

			if err := rows.Scan(&x.SG_Adressen_PK,
				&x.Suchbegriff,
				&x.KundNr,
				&x.LiefNr,
				&x.Homepage,
				&x.Telefon1,
				&x.Telefon2,
				&x.Mobiltelefon1,
				&x.Mobiltelefon2,
				&x.EMail1,
				&x.EMail2,
				&x.KundUmsatz,
				&x.LiefUmsatz); err != nil {
				log.Fatal(err)
				fehler := err.Error()
				json.NewEncoder(w).Encode(map[string]string{"error": fehler})
				return
			}
			SearchResults = append(SearchResults, x)
		}
		if err := rows.Err(); err != nil {
			h.logger.Error("failed to scan rows", slog.Any("error", err))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
	} else {
		rows, err := database.Query(fmt.Sprintf(`
		DECLARE @SearchWord NVARCHAR(30) 
		SET @SearchWord = N'%%%s%%' 
		SELECT 
		SG_Adressen_PK, 
		Suchbegriff,  
		KundNr, 
		LiefNr, 
		Homepage, 
		Telefon1, 
		Telefon2, 
		Mobiltelefon1, 
		Mobiltelefon2, 
		EMail1, 
		EMail2, 
		KundUmsatz, 
		LiefUmsatz 
		FROM sg_adressen 
		WHERE Suchbegriff LIKE @SearchWord 
		OR KundNr LIKE @SearchWord 
		OR LiefNr LIKE @SearchWord;`, searchParams))

		if err != nil {
			h.logger.Error("failed to get sage connection", slog.Any("error", err))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
		defer rows.Close()

		for rows.Next() {
			var x types.Sg_Adressen
			if err := rows.Scan(
				&x.SG_Adressen_PK,
				&x.Suchbegriff,
				&x.KundNr,
				&x.LiefNr,
				&x.Homepage,
				&x.Telefon1,
				&x.Telefon2,
				&x.Mobiltelefon1,
				&x.Mobiltelefon2,
				&x.EMail1,
				&x.EMail2,
				&x.KundUmsatz,
				&x.LiefUmsatz); err != nil {
				log.Fatal(err)
				fehler := err.Error()
				json.NewEncoder(w).Encode(map[string]string{"error": fehler})
				return
			}
			SearchResults = append(SearchResults, x)
		}
		if err := rows.Err(); err != nil {
			h.logger.Error("failed to scan rows", slog.Any("error", err))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
	}

	component.Kunde(SearchResults, searchParams, true).Render(ctx, w)
}
